### DISCUSSION TITLE
일급 함수란 무엇인가요?

### DISCUSSION THREAD

**Discussion by airman5573** [woowacourse-study/2022-woowahan-bansanghwe#106] [JavaScript]
간단히 말하자면, 어떤 함수가 일급 객체와 동일하게 다루어 질때, 일급 함수라고 합니다.  

## 그럼 일급 객체는 무엇인가요?  
일급 객체는 일급 시민인 객체를 의미합니다.  

## 일급 시민은 무엇인가요?  
간단히 말하자면 시민으로서 누릴 수 있는 권한을 모두 누릴 수 있는 사람을 일급 시민이라고 합니다.  

## 어떤 권한이요?  
- 1급 시민은 투표권이 있다
- 1급 시민은 군인이 될 수 있다
- 1급 시민은 정치에 참여 할 수 있다

## 이런게 왜 나온건가요? 2급 시민도 있나요?  
1800년데 영국에서는 1급시민, 2급시민이 존재 했으며  
성인 남성의 경우 1급시민으로서 정치참여가 가능한데 반해,  
**2급시민인 여성은 정치 참여를 할 수 없었다**고 합니다.

## 이게 프로그래밍이랑 무슨 상관이 있나요?  
제 생각에는 값의 성질에 따라서 급을 나눌때 저 시민의 개념을 차용한것이 아닌가 싶습니다.

예를들어, 자바스크립트의 함수는
- 변수에 할당 될 수 있고
- 함수의 인자가 될 수 있고
- 함수의 리턴값이 될 수 있습니다.
그래서 일급 시민의 자격이 있고, 그래서 일급 함수(일급 시민 함수)라고 부를 수 있습니다.

반면, Pascal에서는 함수를 인자로 받을 수 있지만 함수를 리턴할 수는 없습니다.  

## 자바스크립트에서 함수를 일급 객체로 취급하는 이유는 무엇인가요?  
클로저를 사용할 수 있기 때문에, 더 좋습니다!  

이상입니다 :D

**Comment by al-bur** [comment]
자기소개

**Comment by al-bur** [comment]
왜 탐구하는 개발자죠?

**Comment by al-bur** [comment]
작업시간이 얼마 남지 않았는데, 그때도 탐구를 하시나요?

**Comment by al-bur** [comment]
일급 함수란 무엇인가요?

**Comment by al-bur** [comment]
값처럼 쓰이는 함수란 무엇인지 자세히 말해주세요

**Comment by al-bur** [comment]
왜 컴퓨터 과학에서 일급함수가 등장했나요?

**Comment by al-bur** [comment]
포인터는 뭔가요

**Comment by al-bur** [comment]
자바스크립트에도 포인터가 있나요?

**Comment by al-bur** [comment]
참조에 의한 호출

**Comment by al-bur** [comment]
call by reference, call by value 중 자바스크립트의 호출 방식은?

**Comment by al-bur** [comment]
왜 number와 object에는 다른 동작 방식을 갖고 있나요?

**Comment by al-bur** [comment]
heap은 뭔가요?

**Comment by al-bur** [comment]
운영체제 관점에서 크롬을 클릭해서 창이 뜨기까지의 과정

**Comment by al-bur** [comment]
프로세스와 쓰레드의 차이

**Comment by greenblues1190** [comment]
프로세스 간 어떻게 통신할 수 있나요?

**Comment by greenblues1190** [comment]
프로세스가 스레드를 관리한다면 OS에서 프로세스는 누가 관리하나요?
